TEST_APPS/device/nfcapp/nfccommands.cpp

```diff
@@ -0,0 +1,203 @@
+
+
+
+NFCTestShim* new_testshim() {
+#if MBED_CONF_NFCEEPROM
+    mbed::nfc::NFCEEPROMDriver& eeprom_driver = get_eeprom_driver(nfcQueue);
+
+    return ( (NFCTestShim *)(new NFCProcessEEPROM(nfcQueue, eeprom_driver)) );
+#else
+    return( (NFCTestShim *)(new NFCProcessController(nfcQueue)) );
+#endif // EEPROM
+
+}
+
+void nfcRoutine() {
+    nfcQueue.dispatch_forever();
+}
+
+HandleTestCommand::HandleTestCommand() {
+    osStatus status = nfcThread.start(callback(&nfcRoutine) );
+    MBED_ASSERT(status == osOK);
+}
+
+int HandleTestCommand::cmd_set_last_nfc_error(int argc, char *argv[]){
+    if (argc <= 1){
+        cmd_printf("setlastnfcerror() invalid parameter(s)\r\n");
+        return(CMDLINE_RETCODE_INVALID_PARAMETERS);
+    } else {
+        int value = strtol(argv[1], NULL, 10);
+        nfcQueue.call(NFCTestShim::on_set_last_nfc_error, value);
+    }
+    return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+
+int HandleTestCommand::cmd_init_nfc(int argc, char *argv[]){
+
+	if (pNFC_Test_Shim) {
+		cmd_printf("WARN init called again!\r\n"); // only legal here, if eeprom driver stops talking
+	} else {
+	    pNFC_Test_Shim = new_testshim();
+	}
+	nfcQueue.call(pNFC_Test_Shim, &NFCTestShim::on_init);
+	return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+int HandleTestCommand::cmd_read_message(int argc, char *argv[]) {
+    nfcQueue.call(pNFC_Test_Shim, &NFCTestShim::on_read_nfceeprom);
+
+    return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+int HandleTestCommand::cmd_set_smartposter(int argc, char *argv[]) {
+    if (argc <= 1){
+        cmd_printf("setlastnfcerror() invalid parameter(s)\r\n");
+        return(CMDLINE_RETCODE_INVALID_PARAMETERS);
+    } else {
+        // parse arg and queue it up
+        char * uri = (char*)malloc(MAX_URL_LENGTH+1);
+        if (uri) {
+            strncpy(uri, argv[1], MAX_URL_LENGTH);
+            uri[MAX_URL_LENGTH] = '\0';
+            nfcQueue.call(pNFC_Test_Shim, &NFCTestShim::on_set_smartposter, uri); // called thread must free
+        }
+    }
+    return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+// todo: jira IOTPAN-295
+int HandleTestCommand::cmd_erase(int argc, char *argv[]){
+    nfcQueue.call(pNFC_Test_Shim, &NFCTestShim::on_erase);
+    return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+int HandleTestCommand::cmd_write_long_ndef_message(int argc, char *argv[]){
+    size_t length;
+    static const char alphabet[]="thequickbrownfoxjumpedoverthelazydog";
+    char *data;
+
+    if (argc <2) {
+        cmd_printf("supply length of message\r\n");
+        return (CMDLINE_RETCODE_INVALID_PARAMETERS);
+    }
+
+    int converted = sscanf(argv[1], "%d", &length);
+    if (!converted) {
+        cmd_printf("Cannot convert value to int\r\n");
+        return (CMDLINE_RETCODE_INVALID_PARAMETERS);
+    }
+    if (argc >2) {
+        data = (char*)malloc(strlen(argv[2]) + 1);
+        strcpy(data, argv[2]);
+    } else {
+        data = (char*)malloc(strlen(alphabet) + 1);
```

Comment:
what if data is longer than alphabet?
you should malloc the correct amount from the start